models/bots/runner.go

```diff
@@ -0,0 +1,309 @@
+	Repo        *repo_model.Repository `xorm:"-"`
+	Description string                 `xorm:"TEXT"`
+	Base        int                    // 0 native 1 docker 2 virtual machine
+	RepoRange   string                 // glob match which repositories could use this runner
+
+	Token     string `xorm:"-"`
+	TokenHash string `xorm:"UNIQUE"` // sha256 of token
+	TokenSalt string
+	// TokenLastEight string `xorm:"token_last_eight"` // it's unnecessary because we don't find runners by token
+
+	LastOnline timeutil.TimeStamp `xorm:"index"`
+	LastActive timeutil.TimeStamp `xorm:"index"`
+
+	// Store OS and Artch.
+	AgentLabels []string
+	// Store custom labes use defined.
+	CustomLabels []string
+
+	Created timeutil.TimeStamp `xorm:"created"`
+	Updated timeutil.TimeStamp `xorm:"updated"`
+	Deleted timeutil.TimeStamp `xorm:"deleted"`
+}
+
+func (Runner) TableName() string {
+	return "bots_runner"
+}
+
+func (r *Runner) OwnType() string {
+	if r.OwnerID == 0 {
+		return "Global"
+	}
+	if r.RepoID == 0 {
+		return r.Owner.Name
+	}
+
+	return r.Repo.FullName()
+}
+
+func (r *Runner) Status() runnerv1.RunnerStatus {
+	if time.Since(r.LastOnline.AsTime()) > time.Minute {
+		return runnerv1.RunnerStatus_RUNNER_STATUS_OFFLINE
+	}
+	if time.Since(r.LastActive.AsTime()) > 10*time.Second {
+		return runnerv1.RunnerStatus_RUNNER_STATUS_IDLE
+	}
+	return runnerv1.RunnerStatus_RUNNER_STATUS_ACTIVE
+}
+
+func (r *Runner) StatusName() string {
+	return strings.ToLower(strings.TrimPrefix(r.Status().String(), "RUNNER_STATUS_"))
+}
+
+func (r *Runner) IsOnline() bool {
+	status := r.Status()
+	if status == runnerv1.RunnerStatus_RUNNER_STATUS_IDLE || status == runnerv1.RunnerStatus_RUNNER_STATUS_ACTIVE {
+		return true
+	}
+	return false
+}
+
+// AllLabels returns agent and custom labels
+func (r *Runner) AllLabels() []string {
+	return append(r.AgentLabels, r.CustomLabels...)
+}
+
+// Editable checks if the runner is editable by the user
+func (r *Runner) Editable(ownerID, repoID int64) bool {
+	if ownerID == 0 && repoID == 0 {
+		return true
+	}
+	if ownerID > 0 && r.OwnerID == ownerID {
+		return true
+	}
+	return repoID > 0 && r.RepoID == repoID
+}
+
+// LoadAttributes loads the attributes of the runner
+func (r *Runner) LoadAttributes(ctx context.Context) error {
+	if r.OwnerID > 0 {
+		var user user_model.User
+		has, err := db.GetEngine(ctx).ID(r.OwnerID).Get(&user)
+		if err != nil {
+			return err
+		}
+		if has {
+			r.Owner = &user
+		}
+	}
+	if r.RepoID > 0 {
+		var repo repo_model.Repository
+		has, err := db.GetEngine(ctx).ID(r.RepoID).Get(&repo)
+		if err != nil {
+			return err
+		}
+		if has {
+			r.Repo = &repo
+		}
+	}
+	return nil
+}
+
+func (r *Runner) GenerateToken() error {
+	salt, err := util.CryptoRandomString(10)
+	if err != nil {
+		return err
+	}
+	r.TokenSalt = salt
+	r.Token = base.EncodeSha1(gouuid.New().String())
+	r.TokenHash = auth_model.HashToken(r.Token, r.TokenSalt)
+	return nil
+}
```

Comment:
I'm not sure how secure this needs to be, but this doesn't pass my eyeball test either way. 

1. For salt, don't use `CryptoRandomString"; Gitea made that mistake in the past. Instead, use `CryptoRandomBytes(16)` in base64 or hex (or leave it as `[]byte`).
2. Why SHA1 a UUID? The token is then hashed once more for `TokenHash`, which seems excessive to me (not sure what the intentions were). Depending on how secret/secure this needs to be, either use the UUID without hashing or use `CryptoRandomString(16)` or `CryptoRandomBytes(16)` in base64 or hex (or leave it as `[]byte`).

Reply:
The token of a runner or task works like an `auth.AccessToken` somehow, so I tend to have them share logic and format.

1. "For salt, don't use ... Gitea made that mistake in the past", I don't know this, could you give some context? The `auth.AccessToken` does that the same way, see https://github.com/go-gitea/gitea/blob/e81ccc406bf723a5a58d685e7782f281736affd4/models/auth/token.go#L98-L109
2. I think the purpose is formatting, not encrypting, and I cannot answer why not `CryptoRandomString`.

Maybe it's a bit rigid to copy from `auth.AccessToken`, but I want to be conservative when it comes to security, if there's a PR upgraded the logic of `auth.AccessToken`, I'll sync it without any doubt.

Reply:
It's improper to do `hash(uuid)` in any case (what's more that `sha1` is deprecated).

If you need some random tokens, just do fully random byte reading and formatting.

Reply:
Please review #21959, I'll sync it when it has been merged.